go-controller/pkg/ovn/controller/egress_services/egress_services_controller.go

```diff
@@ -430,3 +478,316 @@ func (c *Controller) repair() error {
 
 	return nil
 }
+
+// onEgressServiceAdd queues the EgressService for processing.
+func (c *Controller) onEgressServiceAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	c.egressServiceQueue.Add(key)
+}
+
+// onEgressServiceUpdate queues the EgressService for processing.
+func (c *Controller) onEgressServiceUpdate(oldObj, newObj interface{}) {
+	oldEQ := oldObj.(*egressserviceapi.EgressService)
+	newEQ := newObj.(*egressserviceapi.EgressService)
+
+	if oldEQ.ResourceVersion == newEQ.ResourceVersion ||
+		!newEQ.GetDeletionTimestamp().IsZero() {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(newObj)
+	if err == nil {
+		c.egressServiceQueue.Add(key)
+	}
+}
+
+// onEgressServiceDelete queues the EgressService for processing.
+func (c *Controller) onEgressServiceDelete(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %+v: %v", obj, err))
+		return
+	}
+	c.egressServiceQueue.Add(key)
+}
+
+func (c *Controller) runEgressServiceWorker(wg *sync.WaitGroup) {
+	for c.processNextEgressServiceWorkItem(wg) {
+	}
+}
+
+func (c *Controller) processNextEgressServiceWorkItem(wg *sync.WaitGroup) bool {
+	wg.Add(1)
+	defer wg.Done()
+
+	key, quit := c.egressServiceQueue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.egressServiceQueue.Done(key)
+
+	err := c.syncEgressService(key.(string))
+	if err == nil {
+		c.egressServiceQueue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	if c.egressServiceQueue.NumRequeues(key) < maxRetries {
+		c.egressServiceQueue.AddRateLimited(key)
+		return true
+	}
+
+	c.egressServiceQueue.Forget(key)
+	return true
+}
+
+func (c *Controller) syncEgressService(key string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	startTime := time.Now()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.Infof("Processing sync for Egress Service %s/%s", namespace, name)
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing Egress Service %s/%s : %v", namespace, name, time.Since(startTime))
+	}()
+
+	es, err := c.egressServiceLister.EgressServices(namespace).Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	svc, err := c.serviceLister.Services(namespace).Get(name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	state := c.services[key]
+	if es == nil && state == nil {
```

Comment:
Would the following make sense?
```
if es == nil {
    if state == nil {
        // The EgressService was deleted and wasn't configured before.
        // We delete it from the unallocated service cache just in case.
        delete(c.unallocatedServices, key)
        return nil
    }
    // The service is configured but does no longer have the egress service resource,
    // meaning we should clear all of its resources.
    return c.clearServiceResources(key, state)
}
```
same approach could apply to `svc == nil`

Reply:
done